### DISCUSSION TITLE
이벤트 핸들러 네이밍 건의

### DISCUSSION THREAD

**Discussion by godhyeongman** [eRentronic/eRentronic-Client#57] [아이디어]
## 주제 
통일된 컨벤션 및 가독성, 생산성을 위한 이벤트 핸들러 네이밍 컨벤션 건의

## 상황

상반된 네이밍 아이디어 -> comment를 통해 장점이 많은 쪽으로 결정하기로 건의 발안

dott -> on | 이벤트 | 태그

JS -> 해당 이벤트 동작

**Comment by godhyeongman** [comment]
1. 이벤트 안에서 여러 가지 동작이 이루어 질 경우 이벤트 동작 명은 정확한 내용을 표현하기 어렵다.
ex ) 데이터 검증후 모달 닫기: onClickCloseButton -> 데이터 검증로직 호출, UI제거 로직 호출 -> 이벤트 바인딩과 로직이 분리되어있음
        closeModal -> 데이터 검증후 UI 제거 -> 이벤트 바인딩, 로직을 전부 closeModal이 처리하고 있음

2. 컴포넌트 구현부 에서 이름만 보고도 이벤트 바인딩인지 로직 처리 인지 판단을 할 수 있다.
```
const closeModal -> 모달 열기 이벤트 함수
const openPopUp -> 팝업 여는 로직 처리

---

const onClickModal 
const openPopUP 
```

4. 에러가 생겼을 경우 해당 JSX 구문을 확인 하지 않고 이름만 보고도 판단이 가능하다.
5. 동일한 상황이 가정이라면  데이터를 요청하는 경우도 
get|데이터명 이 아닌 하나의 과정으로 표현해야 통일되 보인다.
ex=> openOrderForm -> 데이터 패칭, 데이터 검증, 오더폼 UI 변경
이방식은 getOrderForm, onClickOrder, changeFormUI로 분리하는게 옳지 않을까?


---

### 컨벤션 혼용 사용에 관한 생각
코드를 읽을때 추상화 수준이 다르면 생각이 복잡해지기 떄문에 추상화 수준을 통일하고 있습니당.
컨벤션 을 통해서도 이런 부분을 통일하는게 가독성에서 더 좋다고 생각합니다.

**Reply by herrakam** [reply]
너무 늦게 달아서 죄송합니다 ㅠㅠ 오늘 컨디션이 별로 안좋았는데 저녁쯤 되니 갑자기 심한 두통이 생겨서 약효 돌기 전에 작성할 수 가 없었어요 ㅠㅠ 지금도 지끈지끈하네요

### 리엑트에서 이벤트 바인딩과 로직을 분리해야 하는가?
이벤트 안에서 여러 가지 동작이 이루어 질 경우 이벤트 동작 명은 정확한 내용을 표현하기 어렵다는 의견은 적극 동의합니다. 저희가 jsx 내에 콜백 형태를 사용하지 않는 방식으로 컨벤션을 정했고, 동작이 2개 이상 들어가면 두 동작을 하나의 함수로 묶는 방법이 더 좋은 선택 같습니다. 이 경우는 저도 두 함수를 묶어주는 함수명을 바인딩 함수를 구현할 때의 네이밍으로 구현할 거 같아요. 
 그런데 동작이 한 개 인 함수에서 이벤트 바인딩과 로직을 굳이 분리해야 하냐는 생각이 듭니다. 이벤트 객체를 사용하는 경우는 반드시 분리를 해야 할 거고,  로직이 비즈니스 로직으로 분리가 될 정도로 중요한 내용이라 숨겨야 할 필요가 있는 경우 또한 동작이 하나여도 분리하는 게 좋은 것 같습니다. 그런데 동작이 한 개 인 경우라면 이벤트 바인딩과 로직을 분리 하지 않아도 될 것 같습니다.   통일성을 갖추기 위해서 바인딩을 담당하는 하나의 함수를 또 작성하는건  비효율적이라고 생각합니다. 굳이 없어도 되는 함수를 또 만드는 것 같아서요.  

### 리엑트는 이벤트 바인딩을 알아서 해결해준다
전 리엑트로 넘어오면서 큰 장점이라고 느꼈던 점중 하나가 이벤트 바인딩 함수를 특별한 경우 아니면 구현하지 않아도 된다는 점이었습니다. 클릭 이벤트에 동작하는 함수를 만들고 싶으면 바닐라 자바스크립트 같은 경우에는 클릭 이벤트를 바인딩 해주고 로직을 처리해주는 함수를 콜백으로 넘겨줘야 했지만, 리엑트는 컴포넌트에서 onclick에 바로 달아주는 것만으로도 동작이 됬으니까요. 바인딩 함수를 항상 구현해주는건 리엑트의 장점을 활용하지 못하고 바닐라 자바스크립트의 단점을 가져오게 되는 방법 같아요. 물론 필요한 경우는 반드시 해줘야 한다고 생각합니다. 

### 추상화 레벨 관련
추상화 관련해서 말씀을 해주셨는데, 전 이벤트 핸들러는 컴포넌트의 하위 단계에서 발생하는 로직이라 매우 낮은 수준의 추상화 레벨인 것 같아요. 그래서 추상화 수준의 오차범위 내의 수준이라고 생각합니다. 하지만  두 네이밍 방법은 추상화 수준이 동일하다고는 볼 수 없다는 점은 동의합니다. 

일단 여기까지 작성해봤습니다. 사실 코딩 스타일에 차이라 큰 문제는 아닌거 같아서 도트가 생각하시는 네이밍을 따라가도 큰 차이는 없을 것 같습니다!! 도트가 선택하시면 전 따라갈게용

**Reply by godhyeongman** [reply]
리액트에서 이벤트 바인딩과 로직을 분리해야하는가?
-> 제가 생각하는건 closeModal 자체가 이벤트에 바인딩 함수가 아닌가? 라는 점입니다. closeModal의 역할이 이벤트를 입력 받고 안에서 로직을 처리하는 것이라면 중요한건 입력을 받는다는 거라 저는 생각 한 것 같습니다.

## 리액트는 이벤트 바인딩을 알아서 해결해준다.

이벤트 바인딩이 복잡하고 조잡스러운걸 리액트가 편하게 구현하는 방법을 제공 한다 생각합니다.
이벤트 바인딩 함수라고 제가 말을 잘 못 한것 같아요.

정정하면 '이벤트 바인딩 함수를 구현해야한다.' -> '사용자 입력' 과 '입력값 처리' 을 분리해야한다. 가 생각이 였던것 같습니다. 근데 하는일이 하나 일떄는 굳이 분리를 해야하냐? 라는 질문에는 안해도 된다 생각합니다. 내부 적으로 처리하는 로직이 한개라면 크게 문제 될 일은 없으니까요. 중요하다 생각한점은 closeModal은 동작하는 '내용'이고 '관심'은 onClickButton이 아닌가? 라는 생각입니다. 

## 추상화 레벨 관련
이부분이 이해가 안가서 혹시 예제 코드 부탁드려도 될까요?

**Reply by herrakam** [reply]
추상화 관련은 예제 코드를 보여드리기보다는 설명을 더 자세히 해드리는게 좋을 것 같아요 `closeModal()` 이 딱 그 예제 코드가 될 것 같아서요. 저희가 page, templete, component 등 여러 단계별로 추상화 수준을 맞추기로 했었는데 사용자의 입력을 받는 위치는 component 내에서도 가장 작은 ui에서 발생하게 된다고 생각해요. 추상화 레벨이 가장 낮은 레벨이기 때문에, 완벽하게 일치할 수는 없겠지만 오차범위 내의 추상화 레벨이 될 것 같다는 생각이었습니다!

**Comment by herrakam** [comment]
추가적으로 컨벤션의 방향성에 대해서 생각을 해봐야겠다는 생각이 들어서 추가로 얘기해볼게요. 컨벤션을 정하면서 통일성이 향상되서 저도 크게 만족할 수 있었습니다. 그런데 오늘 토의를 하면서  자율성도 어느정도 고려를 해봐야 할 것 같다는 생각이 들었어요. 물론 두 상반된 방향 중에 통일성이 더 중요하다는 건 100% 동의합니다!! 하지만 오늘처럼 추가적으로 컨벤션의 대한 토의를 2시간 넘게 진행하는 케이스가 앞으로도  부지기수일 것 같다는 느낌이 들었습니다. 한사람이 작성한 퀄리티가 나올 때 까지 계속해서 토의를 하게 될 거 같은데 컨벤션에 소비하는 비용이 너무 커지는 것 같아요 ㅠㅠ

**Reply by godhyeongman** [reply]
이부분 저도 동의 합니다. 어떻게 개선 하면 좋을까요??  🤔

**Reply by herrakam** [reply]
통일성과 자율성을 구분할 수 있는 기준이 있어야 할 것 같아요! 어떤 게 그 기준이 될지는 좀 더 생각해봐야 할 거 같습니다 ㅠ 어렵네요...
